lib/profileAgents.js

```diff
@@ -0,0 +1,356 @@
+ */
+exports.getAll = async ({account}) => {
+  assert.string(account, 'account');
+
+  const query = {'profileAgent.account': account};
+  const collection = utils.getCollection('profile-profileAgent');
+  const records = await collection.find(query, {_id: 0}).toArray();
+  if(records.length === 0) {
+    return [];
+  }
+  // TODO: Find proper promise-fun library for concurrency
+  const promises = records.map(async record => {
+    const {capabilityAgent, keystoreAgent} = await _getAgents(record);
+    return {
+      ...record,
+      profileAgent: {
+        ...record.profileAgent,
+        capabilityAgent,
+        keystoreAgent
+      }
+    };
+  });
+  return Promise.all(promises);
+};
+
+/**
+ * Get a Profile Agent
+ *
+ * @param {Object} profile the profile associated with the ProfileAgent.
+ * @param {Object} account the account associated with the ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.getByProfile = async ({account, profile}) => {
+  assert.string(account, 'account');
+  assert.string(profile, 'profile');
+
+  const query = {
+    'profileAgent.account': account,
+    'profileAgent.profile': profile
+  };
+  const collection = utils.getCollection('profile-profileAgent');
+  const record = await collection.findOne(query, {_id: 0});
+  if(!record) {
+    const details = {
+      'profileAgent.account': account,
+      'profileAgent.profile': profile,
+      httpStatusCode: 404,
+      public: true
+    };
+    throw new BedrockError(
+      'Profile agent not found.',
+      'NotFoundError', details);
+  }
+  const {capabilityAgent, keystoreAgent} = await _getAgents(record);
+  return {
+    ...record,
+    profileAgent: {
+      ...record.profileAgent,
+      capabilityAgent,
+      keystoreAgent
+    }
+  };
+};
+
+/**
+ * Get a Profile Agent
+ *
+ * @param {Object} profileAgent the updated profile agent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.update = async ({profileAgent: prAgent}) => {
+  const profileAgent = {...prAgent};
+  delete profileAgent.capabilityAgent;
+  delete profileAgent.keystoreAgent;
+  assertProfileAgent(profileAgent);
+
+  // get existing capability set
+  const {id} = profileAgent;
+  const oldRecord = await exports.get({id});
+
+  // ensure sequence number for old set is one less than new set
+  const {profileAgent: oldProfileAgent} = oldRecord;
+  const expectedSequence = oldProfileAgent.sequence + 1;
+
+  if(profileAgent.sequence !== expectedSequence) {
+    throw new BedrockError(
+      'Could not update profile agent; ' +
+      'unexpected sequence number.',
+      'InvalidStateError', {
+        public: true,
+        httpStatusCode: 409,
+        actual: profileAgent.sequence,
+        expected: expectedSequence
+      });
+  }
+
+  // update record
+  const query = {
+    'profileAgent.id': id,
+    'profileAgent.sequence': oldProfileAgent.sequence
```

Comment:
I'm thinking now that this is being included to hedge against simultaneous updates, is that why sequence is being included.

Reply:
Correct.